src/metabase/query_processor/middleware/visualization_settings.clj

```diff
@@ -0,0 +1,72 @@
+(ns metabase.query-processor.middleware.visualization-settings
+  (:require [medley.core :as m]
+            [metabase.models.card :refer [Card]]
+            [metabase.query-processor.store :as qp.store]
+            [metabase.shared.models.visualization-settings :as mb.viz]
+            [toucan.db :as db]))
+
+(defn- normalize-field-settings
+  [id settings]
+  (mb.viz/db->norm-column-settings {(mb.viz/norm->db-column-ref {::mb.viz/field-id id}) settings}))
+
+(defn- update-card-viz-settings
+  "For each field, fetch its settings from the QP store, convert the settings into the normalized form
+  for visualization settings, and then merge in the card-level column settings."
+  [fields column-viz-settings]
+  (map (fn [[_ id _]]
+         (->> id
+              qp.store/field
+              :settings
+              (#(normalize-field-settings id %))
+              (#(m/deep-merge % (select-keys column-viz-settings [{::mb.viz/field-id id}])))))
+   fields))
+
+(defn- update-native-query-viz-settings
+  "For each column, select its viz settings from the column-viz-settings map by its name."
+  [columns column-viz-settings]
+  (map (fn [column] (-> column
+                        :name
+                        (#(select-keys column-viz-settings [{::mb.viz/column-name %}]))))
+       columns))
+
+(defn- get-viz-settings
+  "Pull viz settings from either the query map or the DB"
+  [query]
+  (if-let [card-id (-> query :info :card-id)]
+    ;; Saved card -> fetch viz settings from DB
+    (mb.viz/db->norm (db/select-one-field :visualization_settings Card :id card-id))
+    ;; Unsaved card -> viz settings passed in query
+    (-> query :viz-settings)))
+
+(defn- is-export-context?
+  [context]
+  (contains? #{:json-download :csv-download :xlsx-download} context))
+
+(defn update-viz-settings
+  "Middleware for fetching, processing and ordering a table's visualization settings so that they can be incorporated
+  into an export.
+
+  Card-level visualization settings are either fetched from the DB (for saved cards) or passed from the frontend
+  in the API call (for unsaved cards). These are merged with the base viz settings for each field that are fetched from
+  the QP store (and defined in the data model settings), and are ordered based on the field list from the query.
+
+  For native queries, only viz settings passed from the frontend are incorporated.
+  TODO: native queries only pass viz settings if columns have been hidden or reordered. This needs to be fixed on the frontend."
```

Comment:
I would add some documentation that mentions where this info is added (i.e. under the `:ordered-col-viz-settings` key in metadata) so we don't need to read the code to figure out where it gets added